src/coreclr/nativeaot/Common/src/Internal/Runtime/CompilerHelpers/StartupCodeHelpers.cs

```diff
@@ -218,12 +218,25 @@ internal static unsafe void RunModuleInitializers()
                 IntPtr moduleInitializerSection = RuntimeImports.RhGetModuleSection(typeManager, ReadyToRunSectionType.ModuleInitializerList, out int length);
                 if (moduleInitializerSection != IntPtr.Zero)
                 {
-                    Debug.Assert(length % IntPtr.Size == 0);
-                    var currentCctor = (delegate*<void>*)moduleInitializerSection;
-                    var cctorsEnd = (void*)(moduleInitializerSection + length);
-                    while (currentCctor < cctorsEnd)
+                    var current = (byte*)moduleInitializerSection;
+                    var end = current + length;
+                    while (current < end)
                     {
-                        (*currentCctor++)();
+                        void* cctor;
+
+                        if (Internal.Runtime.MethodTable.SupportsRelativePointers)
```

Comment:
An alternative solution would be to make the section writeable on non-Windows. (The sections with absolute pointers have to be writeable on non-Windows.)